libs/datagrid/src/lib/ag-grid/use-column-sizes.ts

```diff
@@ -0,0 +1,138 @@
+import { immer } from 'zustand/middleware/immer';
+import { useResizeObserver } from '@vegaprotocol/react-helpers';
+
+const STORAGE_KEY = 'vega_columns_sizes_store';
+const COLUMNS_SET_DEBOUNCE_TIME = 300;
+
+export const useColumnSizesStore = create<{
+  sizes: Record<string, Record<string, number>>;
+  valueSetter: (id: string, value: Record<string, number>) => void;
+}>()(
+  persist(
+    immer((set) => ({
+      sizes: {},
+      valueSetter: (id, value) =>
+        set((state) => {
+          state.sizes[id] = {
+            ...(state.sizes[id] || {}),
+            ...value,
+          };
+          return state;
+        }),
+    })),
+    { name: STORAGE_KEY }
+  )
+);
+
+interface UseColumnSizesProps {
+  id?: string;
+  container?: MutableRefObject<HTMLDivElement | null>;
+}
+export const useColumnSizes = ({
+  id = '',
+  container,
+}: UseColumnSizesProps): [
+  (columns: Column[]) => void,
+  (children?: ReactElement[]) => ReactElement[] | undefined,
+  (
+    props: AgGridReactProps | AgReactUiProps,
+    children?: ReactElement[]
+  ) => AgGridReactProps | AgReactUiProps
+] => {
+  const sizes = useColumnSizesStore((store) => store.sizes[id] || {});
+  const valueSetter = useColumnSizesStore((store) => store.valueSetter);
+  const getWidthOfAll = useCallback(
+    () =>
+      (container?.current as HTMLDivElement)?.getBoundingClientRect().width ??
+      0,
+    [container]
+  );
+  const recalculateSizes = useCallback(
+    (sizes: Record<string, number>) => {
+      const width = getWidthOfAll();
+      if (width && sizes['width'] && width !== sizes['width']) {
+        const oldWidth = sizes['width'];
+        const ratio = width / oldWidth;
+        return {
+          ...Object.entries(sizes).reduce((agg, [key, value]) => {
+            agg[key] = value * ratio;
+            return agg;
+          }, {} as Record<string, number>),
+          width,
+        } as Record<string, number>;
+      }
+      return sizes;
+    },
+    [getWidthOfAll]
+  );
+  const [calculatedSizes, setCalculatedSizes] = useState(
+    recalculateSizes(sizes)
+  );
+  const onResize = useCallback(() => {
+    const width = getWidthOfAll();
+    if (width && sizes['width'] && width !== sizes['width']) {
+      setCalculatedSizes(recalculateSizes(sizes));
+    }
+  }, [getWidthOfAll, recalculateSizes, sizes]);
+  useResizeObserver(container?.current as Element, onResize);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const handleOnChange = useCallback(
+    debounce((columns: Column[]) => {
+      if (id && columns.length) {
+        const sizesObj = columns.reduce((aggr, column) => {
+          aggr[column.getColId()] = column.getActualWidth();
+          return aggr;
+        }, {} as Record<string, number>);
+        sizesObj['width'] = getWidthOfAll();
+        valueSetter(id, sizesObj);
+      }
+    }, COLUMNS_SET_DEBOUNCE_TIME),
+    [valueSetter, id]
+  );
+  const reshapeAgGridChildren = useCallback(
```

Comment:
chore (blocking): I think we can make this simpler by doing the resizing using the api in the `onGridReady` callback, then we don't need to modify children and coldefs.

Additionally we can call `sizeColumnsToFit` if not stored sizes are found so that the columns fill available space. Right now there is unused space on some tables

Something like this should work 🤔 

```
onGridReady={(event) => {
  if (!sizes) {
    event.columnApi.sizeColumnsToFit();
  } else {
    // convert sizes to an array of column size in this shape: { key: "the-col-id", size: number }
    const sizes = Object.entries(sizes).map(
      ([key, size]) => ({
        key,
          newWidth: size,
        })
      );
      event.columnApi.setColumnWidths(sizes);
    }
}}
```

Reply:
I'm afraid, relying on `onGridReady` can bring us blink effect, since it's fired usually after some time

Reply:
ok, after some checks, it doesn't look bad, I mean `setColumnWidths` instead iterating children and props. Please take a look.